### PR TITLE
NodesTreeWidget: _find_outputs parse nested namespace outputs

### LINKED ISSUE #262: NodesTreeWidget: find outputs not working for nested namespace
I have the following outputs with some output nodes collected in the namespace:

```
Outputs                         PK    Type
------------------------------  ----  ------
convergence_cohesive_energy
    output_parameters           8916  Dict
delta_factor
    output_birch_murnaghan_fit  8937  Dict
    output_delta_factor         8943  Dict
    output_volume_energy        8935  Dict
pseudo_info                     7795  Dict
```

It will cause an issue when I want to represent the output as a NodesTreeWidget.
The exceptions is `AttributeError: 'AttributeDict' object has no attribute 'pk'`. 
It is because the `_find_outputs` class method did not detect when the output is a namespace the type `AttributeDict`.

### PR DESCRIPTION
**Pull request by unkcpz**
fix #262

Previous PR #269 only works when the outputs have a single layer of
the nested namespace. We store the `parent_pk` in order to transversal nodes
level by level, however only with extra stored namespace parameter is not enough to access the
outputs in deep level. To solve this, `namespace` replace by `namespaces` which is a 
tuple to store the full path so that all the output nodes can be visited.

### CODE REVIEW

File: aiidalab_widgets_base/nodes.py
Comment (csadorf) [inline aiidalab_widgets_base/nodes.py:225]: I find this segment harder to understand now. Would the following refactoring be correct?
```suggestion
        # Gather outputs from node and its namespaces (if present):
        outputs = functools.reduce(
            lambda d, namespace: d[namespace], root.namespaces or [], process_node.outputs
        )
```

File: aiidalab_widgets_base/nodes.py
Comment (csadorf) [inline aiidalab_widgets_base/nodes.py:227]:
```suggestion
        # Convert aiida.orm.LinkManager to dict
```

File: aiidalab_widgets_base/nodes.py
Comment (csadorf) [inline aiidalab_widgets_base/nodes.py:239]: In combination with above suggestion:
```suggestion
                    namespaces=(*root.namespaces, key),
```

**Review by csadorf — COMMENTED** [review]
This looks sound to me, but I think I would prefer to have a better understanding of the namespaces. Can you point me to the relevant section in the AiiDA documentation?

**Comment by unkcpz** [comment]
@csadorf Thanks a lot! Indeed, the code is more clear with your suggestion, updated.

> Can you point me to the relevant section in the AiiDA documentation?

The introduction of `namespace`  is in the process part of documentation https://aiida.readthedocs.io/projects/aiida-core/en/latest/topics/processes/usage.html?highlight=namespace#ports-and-port-namespaces 

It is then more detail described in the 'Exposing output and input' part https://aiida.readthedocs.io/projects/aiida-core/en/latest/howto/write_workflows.html?highlight=namespace#exposing-inputs-and-outputs since this is the most common way plugin developers use namespace concept implicitly, such as in the `QeWorkChain` in aiidalab-qe.

File: aiidalab_widgets_base/nodes.py
Comment (csadorf) [inline aiidalab_widgets_base/nodes.py:227]: What would happen if namespaces have outputs with the same key?

File: aiidalab_widgets_base/nodes.py
Comment (unkcpz) [reply aiidalab_widgets_base/nodes.py:241]: I don't think this change is correct. The `else` branch is for the condition that the node is not a namespace but a true data node. While the `if node.pk not in root.nodes_registry` is for the caching tree nodes.

File: aiidalab_widgets_base/nodes.py
Comment (unkcpz) [reply aiidalab_widgets_base/nodes.py:227]: This should already be validated by aiida, it is not allowed to have duplicate keys under one namespace.

File: aiidalab_widgets_base/nodes.py
Comment (csadorf) [reply aiidalab_widgets_base/nodes.py:241]: Yes, sorry about that.

File: aiidalab_widgets_base/nodes.py
Comment (unkcpz) [reply aiidalab_widgets_base/nodes.py:227]: >  there can be no situation where outputs from different namespaces could share the same key ?

For the different namespaces, the keys can be the same. But the 'functools.reduce' unfold the namespace to the innermost nested specific namespace. For one namespace, keys can't be the same.

File: aiidalab_widgets_base/nodes.py
Comment (unkcpz) [reply aiidalab_widgets_base/nodes.py:224]: No worries, I admit this is kind of confused, every time I look at it again I need to go through all the code and try to clean my mind.

Assuming we have the following outputs of a process node `wc_node`:
```
ns00
    ns10
        outnode_a
        outnode_b
ns01
        outnode_a
        outnode_c
```
If we want to access `outnode_a` node, it should be `wc_node.outputs['ns00']['ns01']['outnode_a']`. In the scope (iteration), the `root.namespaces` in this layer equal to `[`ns00`, `ns01`]`.

> Either way, this function probably requires a bit more commenting since it is somewhat intransparent.
Indeed, I will add it if you think the code and explanation above are clear enough.

Comment (unkcpz) [reply aiidalab_widgets_base/nodes.py:224]: `d[namespace]` in the `reduce` make sure that the `d` accumulate the change for every nested `namespace`.

Comment (csadorf) [reply aiidalab_widgets_base/nodes.py:224]: From offline discussion: The code appears to be correct, however it will require a bit more documentation to make it easier to follow. After that I'm happy to approve and merge.

Comment (unkcpz) [reply aiidalab_widgets_base/nodes.py]: Hi @csadorf, I add docstring and comment to make (try to🤣) it more clear of the logic. Please have a look.

File: aiidalab_widgets_base/nodes.py
Comment (csadorf) [inline aiidalab_widgets_base/nodes.py:221]:
```suggestion
        A generator for all (including nested) output nodes.
        
        Generates an AiidaOutputsTreeNode when encountering a namespace,
        keeping track of the full namespace path to make it accessible via the
        root node in form of a breadth-first search.
```

**Review by csadorf — APPROVED** [review]
One suggestion to reformulate the doc-string a bit, but otherwise good to merge. Thanks a lot! 👍

**Comment by unkcpz** [comment]
@csadorf Thanks! Docstring updated and ready to be merged 😄 please approve once more.

**Review by csadorf — APPROVED** [review]

**Pull request merged**
